Provider.zip combiner function cannot return null

<!--- 
Please follow the instructions below. 
We receive dozens of issues every week, so to stay productive, we will close issues that don't provide enough information. 

Please open Android-related issues on the Android Issue Tracker at https://source.android.com/source/report-bugs
Please open Gradle Native-related issues at https://github.com/gradle/gradle-native/issues
-->

<!--- Provide a brief summary of the issue in the title above -->

### Expected Behavior
The combiner function in `Property#zip(a, b, combiner)` returning null should result in the zipped property having a missing value. This matches the behavior from ProviderFactory#provider and Provider#map.

### Current Behavior
Returning `null` from the combiner function throws an `IllegalArgumentException`:

```
Caused by: java.lang.IllegalArgumentException
    at org.gradle.api.internal.provider.ValueSupplier$Value.of(ValueSupplier.java:215)
    at org.gradle.api.internal.provider.BiProvider.calculateOwnValue(BiProvider.java:56)
    at org.gradle.api.internal.provider.AbstractMinimalProvider.calculatePresence(AbstractMinimalProvider.java:79)
    at org.gradle.api.internal.provider.MappingProvider.calculatePresence(MappingProvider.java:50)
    at org.gradle.api.internal.provider.AbstractProperty.calculatePresence(AbstractProperty.java:58)
```

### Context
<!--- How has this issue affected you? What are you trying to accomplish? What other alternatives have you considered? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

I currently have a plugin configuring the CompileJava task like so:

```java
options.getRelease().set(bytecodeVersion.zip(toolchainVersion, (bytecode, toolchain) -> {
        if (toolchain >= 9) {
          return bytecode;
        } else {
          return null;
        }
      }));
```
which fails while returning null and have used this as a temporary workaround (even if octy golfed something a bit nicer :p):

```java
      final Provider<Integer> releaseFlagToolchain = toolchainVersion.map(v -> {
        if (v > 9) {
          return v;
        } else {
          return null;
        }
      });

      options.getRelease().set(bytecodeVersion.zip(releaseFlagToolchain, (bytecode, toolchain) -> bytecode));
```

--- a/subprojects/core-api/src/main/java/org/gradle/api/provider/Provider.java
+++ b/subprojects/core-api/src/main/java/org/gradle/api/provider/Provider.java
@@ -224,11 +224,11 @@ public interface Provider<T> {
      * </p>
      *
      * @param right the second provider to combine with
-     * @param combiner the combiner of values
+     * @param combiner the combiner of values. May return {@code null}, in which case the provider
+     * will have no value.
      * @param <U> the type of the second provider
      * @param <R> the type of the result of the combiner
      * @return a combined provider
-     *
      * @since 6.6
      */
     <U, R> Provider<R> zip(Provider<U> right, BiFunction<? super T, ? super U, ? extends R> combiner);

--- a/subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderFactory.java
+++ b/subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderFactory.java
@@ -295,17 +295,19 @@ public interface ProviderFactory {
      * Returns a provider which value will be computed by combining a provider value with another
      * provider value using the supplied combiner function.
      *
+     * <p>
      * If the supplied providers represents a task or the output of a task, the resulting provider
      * will carry the dependency information.
+     * </p>
      *
      * @param first the first provider to combine with
      * @param second the second provider to combine with
-     * @param combiner the combiner of values
+     * @param combiner the combiner of values. May return {@code null}, in which case the provider
+     * will have no value.
      * @param <A> the type of the first provider
      * @param <B> the type of the second provider
      * @param <R> the type of the result of the combiner
      * @return a combined provider
-     *
      * @since 6.6
      */
     <A, B, R> Provider<R> zip(Provider<A> first, Provider<B> second, BiFunction<? super A, ? super B, ? extends R> combiner);

--- a/subprojects/core/src/test/groovy/org/gradle/api/internal/file/DefaultProjectLayoutTest.groovy
+++ b/subprojects/core/src/test/groovy/org/gradle/api/internal/file/DefaultProjectLayoutTest.groovy
@@ -177,7 +177,7 @@ class DefaultProjectLayoutTest extends Specification {
     }
 
     def "can resolve directory relative to build directory"() {
-        def pathProvider = withValues("a", "b")
+        def pathProvider = withValues("a", "b", "c")
 
         expect:
         def dir = layout.buildDirectory.dir("sub-dir")
@@ -188,17 +188,17 @@ class DefaultProjectLayoutTest extends Specification {
         provider.present
 
         def dir1 = provider.get()
-        dir1.getAsFile() == projectDir.file("build/a")
+        dir1.getAsFile() == projectDir.file("build/b")
 
         def dir2 = provider.get()
-        dir2.getAsFile() == projectDir.file("build/b")
+        dir2.getAsFile() == projectDir.file("build/c")
 
-        dir1.getAsFile() == projectDir.file("build/a")
-        dir2.getAsFile() == projectDir.file("build/b")
+        dir1.getAsFile() == projectDir.file("build/b")
+        dir2.getAsFile() == projectDir.file("build/c")
     }
 
     def "can resolve regular file relative to build directory"() {
-        def pathProvider = withValues("a", "b")
+        def pathProvider = withValues("a", "b", "c")
 
         expect:
         def file = layout.buildDirectory.file("child")
@@ -209,13 +209,13 @@ class DefaultProjectLayoutTest extends Specification {
         provider.present
 
         def file1 = provider.get()
-        file1.getAsFile() == projectDir.file("build/a")
+        file1.getAsFile() == projectDir.file("build/b")
 
         def file2 = provider.get()
-        file2.getAsFile() == projectDir.file("build/b")
+        file2.getAsFile() == projectDir.file("build/c")
 
-        file1.getAsFile() == projectDir.file("build/a")
-        file2.getAsFile() == projectDir.file("build/b")
+        file1.getAsFile() == projectDir.file("build/b")
+        file2.getAsFile() == projectDir.file("build/c")
     }
 
     def "directories are equal when their paths are equal"() {

--- a/subprojects/model-core/src/main/java/org/gradle/api/internal/provider/BiProvider.java
+++ b/subprojects/model-core/src/main/java/org/gradle/api/internal/provider/BiProvider.java
@@ -16,6 +16,7 @@
 package org.gradle.api.internal.provider;
 
 import org.gradle.api.provider.Provider;
+import org.jetbrains.annotations.NotNull;
 
 import javax.annotation.Nullable;
 import java.util.function.BiFunction;
@@ -41,7 +42,11 @@ public class BiProvider<R, A, B> extends AbstractMinimalProvider<R> {
 
     @Override
     public boolean calculatePresence(ValueConsumer consumer) {
-        return left.calculatePresence(consumer) && right.calculatePresence(consumer);
+        if (!left.calculatePresence(consumer) || !right.calculatePresence(consumer)) {
+            return false;
+        }
+        // Purposefully only calculate full value if left & right are both present, to save time
+        return !calculateValue(consumer, false).isMissing();
     }
 
     @Override
@@ -57,6 +62,11 @@ public class BiProvider<R, A, B> extends AbstractMinimalProvider<R> {
 
     @Override
     protected Value<? extends R> calculateOwnValue(ValueConsumer consumer) {
+        return calculateValue(consumer, true);
+    }
+
+    @NotNull
+    private Value<R> calculateValue(ValueConsumer consumer, boolean keepSideEffects) {
         Value<? extends A> leftValue = left.calculateValue(consumer);
         if (leftValue.isMissing()) {
             return leftValue.asType();
@@ -68,9 +78,13 @@ public class BiProvider<R, A, B> extends AbstractMinimalProvider<R> {
 
         R combinedUnpackedValue = combiner.apply(leftValue.getWithoutSideEffect(), rightValue.getWithoutSideEffect());
 
-        return Value.of(combinedUnpackedValue)
-            .withSideEffect(SideEffect.fixedFrom(leftValue))
-            .withSideEffect(SideEffect.fixedFrom(rightValue));
+        Value<R> value = Value.ofNullable(combinedUnpackedValue);
+        if (keepSideEffects && !value.isMissing()) {
+            value = value
+                .withSideEffect(SideEffect.fixedFrom(leftValue))
+                .withSideEffect(SideEffect.fixedFrom(rightValue));
+        }
+        return value;
     }
 
     @Nullable

--- a/subprojects/model-core/src/main/java/org/gradle/api/internal/provider/BiProvider.java
+++ b/subprojects/model-core/src/main/java/org/gradle/api/internal/provider/BiProvider.java
@@ -16,7 +16,6 @@
 package org.gradle.api.internal.provider;
 
 import org.gradle.api.provider.Provider;
-import org.jetbrains.annotations.NotNull;
 
 import javax.annotation.Nullable;
 import java.util.function.BiFunction;
@@ -46,7 +45,7 @@ public class BiProvider<R, A, B> extends AbstractMinimalProvider<R> {
             return false;
         }
         // Purposefully only calculate full value if left & right are both present, to save time
-        return !calculateValue(consumer, false).isMissing();
+        return super.calculatePresence(consumer);
     }
 
     @Override
@@ -62,11 +61,6 @@ public class BiProvider<R, A, B> extends AbstractMinimalProvider<R> {
 
     @Override
     protected Value<? extends R> calculateOwnValue(ValueConsumer consumer) {
-        return calculateValue(consumer, true);
-    }
-
-    @NotNull
-    private Value<R> calculateValue(ValueConsumer consumer, boolean keepSideEffects) {
         Value<? extends A> leftValue = left.calculateValue(consumer);
         if (leftValue.isMissing()) {
             return leftValue.asType();
@@ -78,13 +72,9 @@ public class BiProvider<R, A, B> extends AbstractMinimalProvider<R> {
 
         R combinedUnpackedValue = combiner.apply(leftValue.getWithoutSideEffect(), rightValue.getWithoutSideEffect());
 
-        Value<R> value = Value.ofNullable(combinedUnpackedValue);
-        if (keepSideEffects && !value.isMissing()) {
-            value = value
-                .withSideEffect(SideEffect.fixedFrom(leftValue))
-                .withSideEffect(SideEffect.fixedFrom(rightValue));
-        }
-        return value;
+        return Value.ofNullable(combinedUnpackedValue)
+            .withSideEffect(SideEffect.fixedFrom(leftValue))
+            .withSideEffect(SideEffect.fixedFrom(rightValue));
     }
 
     @Nullable

--- a/subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/DefaultProviderFactoryTest.groovy
+++ b/subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/DefaultProviderFactoryTest.groovy
@@ -112,6 +112,7 @@ class DefaultProviderFactoryTest extends Specification implements ProviderAssert
 
         then:
         !zipped.isPresent()
+        zipped.getOrNull() == null
     }
 
     def "can zip two providers of arbitrary types"() {

--- a/subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/DefaultProviderFactoryTest.groovy
+++ b/subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/DefaultProviderFactoryTest.groovy
@@ -21,8 +21,6 @@ import org.gradle.api.provider.Provider
 import org.gradle.testfixtures.ProjectBuilder
 import spock.lang.Specification
 
-import java.util.concurrent.atomic.AtomicReference
-
 import static org.gradle.api.internal.provider.ProviderTestUtil.withProducer
 import static org.gradle.api.internal.provider.ProviderTestUtil.withValues
 
@@ -93,22 +91,18 @@ class DefaultProviderFactoryTest extends Specification implements ProviderAssert
     }
 
     def "can zip two providers and use null to remove the value, and it is live"() {
-        def reference = new AtomicReference<String>("initial")
-        def provider = providerFactory.provider(reference::get)
+        def provider = withValues("accepted", "accepted", "rejected", "rejected")
         def second = providerFactory.provider { "value" }
 
         when:
         def zipped = providerFactory.zip(provider, second) { s1, s2 ->
-            s1 == "initial" ? "$s1 $s2" : null
+            s1 == "accepted" ? "$s1 $s2" : null
         }
 
         then:
         zipped instanceof Provider
         zipped.isPresent()
-        zipped.get() == "initial value"
-
-        when:
-        reference.set("changed")
+        zipped.get() == "accepted value"
 
         then:
         !zipped.isPresent()

--- a/subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/DefaultProviderFactoryTest.groovy
+++ b/subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/DefaultProviderFactoryTest.groovy
@@ -21,6 +21,8 @@ import org.gradle.api.provider.Provider
 import org.gradle.testfixtures.ProjectBuilder
 import spock.lang.Specification
 
+import java.util.concurrent.atomic.AtomicReference
+
 import static org.gradle.api.internal.provider.ProviderTestUtil.withProducer
 import static org.gradle.api.internal.provider.ProviderTestUtil.withValues
 
@@ -75,6 +77,41 @@ class DefaultProviderFactoryTest extends Specification implements ProviderAssert
         then:
         zipped instanceof Provider
         zipped.get() == 'Big black cat'
+    }
+
+    def "can zip two providers and use null to remove the value"() {
+        def not = providerFactory.provider { "not" }
+        def important = providerFactory.provider { "important" }
+
+        when:
+        def zipped = providerFactory.zip(not, important) { s1, s2 -> null }
+
+        then:
+        zipped instanceof Provider
+        !zipped.isPresent()
+        zipped.getOrNull() == null
+    }
+
+    def "can zip two providers and use null to remove the value, and it is live"() {
+        def reference = new AtomicReference<String>("initial")
+        def provider = providerFactory.provider(reference::get)
+        def second = providerFactory.provider { "value" }
+
+        when:
+        def zipped = providerFactory.zip(provider, second) { s1, s2 ->
+            s1 == "initial" ? "$s1 $s2" : null
+        }
+
+        then:
+        zipped instanceof Provider
+        zipped.isPresent()
+        zipped.get() == "initial value"
+
+        when:
+        reference.set("changed")
+
+        then:
+        !zipped.isPresent()
     }
 
     def "can zip two providers of arbitrary types"() {

--- a/subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/WithSideEffectProviderTest.groovy
+++ b/subprojects/model-core/src/test/groovy/org/gradle/api/internal/provider/WithSideEffectProviderTest.groovy
@@ -186,7 +186,9 @@ class WithSideEffectProviderTest extends ProviderSpec<Integer> {
         def zippedSideEffect = Mock(ValueSupplier.SideEffect)
         def leftWithSideEffect = Providers.ofNullable(leftValue).withSideEffect(leftSideEffect)
         def rightWithSideEffect = Providers.ofNullable(rightValue).withSideEffect(rightSideEffect)
-        def zipped = leftWithSideEffect.zip(rightWithSideEffect) { a, b -> a + b } as ProviderInternal<Integer>
+        def zipped = leftWithSideEffect.zip(rightWithSideEffect) { a, b ->
+            a == Integer.MAX_VALUE ? null : (a + b)
+        } as ProviderInternal<Integer>
         def provider = zipped.withSideEffect(zippedSideEffect)
 
         when:
@@ -211,11 +213,12 @@ class WithSideEffectProviderTest extends ProviderSpec<Integer> {
         0 * _
 
         where:
-        description      | leftValue | leftRuns | rightValue | rightRuns | expectedResult
-        "both have"      | 23        | 1        | 88         | 1         | 23 + 88
-        "only left has"  | 23        | 0        | null       | 0         | null
-        "only right has" | null      | 0        | 88         | 0         | null
-        "none have"      | null      | 0        | null       | 0         | null
+        description                            | leftValue         | leftRuns | rightValue | rightRuns | expectedResult
+        "both have"                            | 23                | 1        | 88         | 1         | 23 + 88
+        "both have (but zip provider missing)" | Integer.MAX_VALUE | 0        | 88         | 0         | null
+        "only left has"                        | 23                | 0        | null       | 0         | null
+        "only right has"                       | null              | 0        | 88         | 0         | null
+        "none have"                            | null              | 0        | null       | 0         | null
     }
 
     def "carries the side effect in the execution time value for a provider of a fixed value"() {

